compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -48,6 +52,42 @@ use rustc_hir::intravisit::Visitor;
 use std::cmp::{self, Ordering};
 use std::iter;
 
+#[derive(Debug)]
+struct ConfusablesDirective {
+    names: Vec<String>,
+}
+
+#[derive(Diagnostic)]
+#[diag(hir_typeck_invalid_rustc_confusable_attr)]
+struct ConfusablesDirectiveParseErr {
+    #[primary_span]
+    #[label]
+    span: Span,
+}
+
+impl ConfusablesDirective {
+    fn try_parse(attr: &NormalAttr) -> Result<Self, ConfusablesDirectiveParseErr> {
```

Comment:
```suggestion
    fn try_parse(attr: &NormalAttr) -> Result<ConfusablesDirective, ConfusablesDirectiveParseErr> {
```
Nit: `Self` is better kept for generic contexts.